[FLINK-16615] Introduce data structures and utilities to calculate Job Manager memory components

azagrebin opened the pull request:
## What is the purpose of the change

This PR introduces new memory model and its configuration options for JM.
There is no behavioural change at the moment, only data structures for memory components are prepared and options are parsed to fill them.
See more details in [FLIP-116: Unified Memory Configuration for Job Managers](https://cwiki.apache.org/confluence/display/FLINK/FLIP+116%3A+Unified+Memory+Configuration+for+Job+Managers).

## Brief change log

  - Introduce new memory JM options in `JobManagerOptions`
  - Introduce `JobManagerProcessSpec` container for JM memory components
  - Factor out common parts from `TaskExecutorProcessUtils` into `MemoryProcessUtils` which can be reused for JM (JVM components and arguments, some other common methods)
  - Factor out common tests from `TaskExecutorProcessUtilsTest` into the base test suite `MemoryProcessUtilsTestBase`
 - Introduce `JobManagerProcessUtils` which parses new memory specific `JobManagerOptions` and calculates JM memory components as `JobManagerProcessSpec`
 - Introduce `JobManagerProcessUtilsTest` suite for `JobManagerProcessUtils`

## Verifying this change

Unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable now)

azagrebin commented:
cc @xintongsong

azagrebin commented:
Thanks for the review @xintongsong 
I refactored more the class structure. The final composition is a bit different but goes into a similar direction. This should also allow better test separation (probably later) comparing to the inheritance approach which is also in one of commits.
I also addressed smaller comments.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/LegacyHeapOptions.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Legacy JVM heap/process memory options to fallback to new options.
+ */
+public class LegacyHeapOptions {
```
How about name this `ProcessMemoryFallbackOptions` or `ProcessMemoryLegacyOptions`?
I think "heap" could be ambiguous because it's used both in the old configuration as the total memory, and in the new configuration as the java heap memory. To that end, "process" might be a clearer phrase.

azagrebin replied:
True, I would suggest then just `LegacyMemoryOptions` with more details in jdoc comment.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/LegacyMemoryUtils.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Utilities to fallback to new options from the lecacy ones.
+ */
+public class LegacyMemoryUtils {
```
How about `ProcessMemoryBackwardsCompatibilityUtils`?

azagrebin replied:
ok, I would again suggest `MemoryBackwardsCompatibilityUtils` for brevity and that it is sometimes about heap, sometimes about total process memory.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtils.java:
```diff
@@ -0,0 +1,117 @@
+		JobManagerFlinkMemory flinkMemory = processMemory.getFlinkMemory();
+		JvmMetaspaceAndOverhead jvmMetaspaceAndOverhead = processMemory.getJvmMetaspaceAndOverhead();
+		verifyJvmHeapSize(flinkMemory.getJvmHeapMemorySize());
+		verifyJobStoreCacheSize(config, flinkMemory.getJvmHeapMemorySize());
```
Minor: Should these two verifications could be inside `JobManagerFlinkMemory`?

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemorySpecBase.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.flink.configuration.MemorySize;
+
+/**
+ * Process memory components.
```
I would suggest:
`Common memory components of Flink processes (e.g. JM or TM).`
also, maybe, `CommonProcessMemorySpec` is better name because it is self-sufficient on its and not really always abstract (e.g. returned from `ProcessMemoryUtils`). We just choose for now to extend it for JM/TM. I can see it also being a field of some specs, maybe in future making it a composition.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemorySpecBase.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Process memory components.
+ */
+public class ProcessMemorySpecBase<IM extends FlinkMemory> implements ProcessMemorySpec {
```
see other comment

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java:
```diff
@@ -0,0 +1,273 @@
+public class ProcessMemoryUtils<IM extends FlinkMemory> {
+	private static final Logger LOG = LoggerFactory.getLogger(ProcessMemoryUtils.class);
+
+	private final List<ConfigOption<MemorySize>> requiredFineGrainedOptions;
```
True, although,  I would suggest to use rename
`FlinkMemoryUtils#deriveFromInternalMemory -> FlinkMemoryUtils#deriveFromFineGrainedOptions`
I think `internal` was more confusing to me

azagrebin commented:
Thanks for more review @xintongsong 
I addressed the comments.
The PR is already big. It makes sense to improve tests as a follow-up. I will create an issue soon.

Jiayi-Liao commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/ProcessMemoryUtils.java:
```diff
@@ -0,0 +1,281 @@
+	}
+
+	public static String generateJvmParametersStr(ProcessMemorySpec processSpec) {
+		return "-Xmx" + processSpec.getJvmHeapMemorySize().getBytes()
```
@azagrebin @xintongsong 

Hi, I'm upgrading our Flink to 1.11 version. And I find that the memory related arguments in JVM parameters are hard to read here because we're using bytes here, for example:

```
jdk/jdk1.8/bin/java -Xmx944892739 -Xms944892739 -XX:MaxDirectMemorySize=1213328302
```

I wonder whether we can make some minor changes here to improve this. Or did you use bytes on purpose for other reasons I probably missed?

xintongsong replied:
@Jiayi-Liao,

I believe these parameters are intended to use bytes, in order to preserve the accuracy. We have run into troubles previously caused by rounding errors when using megabytes. So we decided to always use bytes for Flink's internal memory setups and calculations.

I agree with you that it would be helpful to have some human readable representations for the memory sizes and JVM parameters. But I think the parameters here are intended to be read by the JVM rather than users. How about we print such information in the beginning of the logs?

@azagrebin 

I think we probably should also log the parameters/configurations in a human readable way in `ProcessMemoryUtils.generateJvmParametersStr` and `TaskExecutorProcessUtils.generateDynamicConfigsStr`. WDYT?

azagrebin replied:
@xintongsong @Jiayi-Liao 
I agree we can log the human readable values for JVM memory arguments as they are not always just option values.
We can also do it for  `generateDynamicConfigsStr` on debug level, the dynamic properties will be also outputted in startup logs of the TM JVM process itself.

xintongsong replied:
FYI, I've created FLINK-18175 to track this.
